File: trapping-rain-water.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int trap(vector<int>& height) {
        int l = 0;
        int r = height.size() - 1;
        int left_max = INT_MIN;
        int right_max = INT_MIN;
        int water = 0;
        while(l < r){
            left_max = max(left_max, height[l]);
            right_max = max(right_max, height[r]);
            water += left_max < right_max ? left_max - height[l++] : right_max - height[r--];

        }
        return water;
    }

};

File: trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
       left = 0
       right = len(height) - 1
       water = 0
       left_max = float('-inf')
       right_max = float('-inf')
       while left < right: 
           left_max = max(left_max, height[left])
           right_max = max(right_max, height[right])
           if left_max < right_max: 
               water += left_max - height[left]
               left += 1
           else:
                water += right_max - height[right]
                right -= 1       
       return water





        